### PR TITLE
Implement File Format Reader/Writer

### LINKED ISSUE #54: Implement File Format Reader/Writer
# Description

Implement the initial methods to read and write `.csv`, `.json`, `.orc`, `.parquet`, and `.text`. 

Consider creating `ConfigOpts` trait for each of those file options and have a custom struct represent the options for each of those file types. 

## Example with CSV

Create the Options and modify the opts object. The object is passed into the method.

```rust

let mut opts = CsvOptions::new()

opts.header = true;
opts.delimiter = b'|';

let df = spark.read().csv(path, opts)
```

Example of what the function signature might look like

```rust
impl DataFrameReader {
 ....
     pub fn csv<C: ConfigOpts>(path: &str, opts: Some(C))
}

### PR DESCRIPTION
**Pull request by lexara-prime-ai**
# Description

This pull request enables DataFrames to be **read from and written to various file formats** (CSV, JSON, ORC, Parquet) using a **set of predefined options** implemented via the `ConfigOpts` trait. Key changes include:

-   **DataFrameWriter:**
    
    -   Implemented functionality to write DataFrames to `csv`, `json`, `orc`, `parquet` and `text` formats.
    -   Added support for configuration options for various file formats (i.e `csv`, `json`, `orc`, `parquet` and `text`).
    -   `ConfigOpts` trait implemented for `JsonOptions`, `OrcOptions`, and `ParquetOptions`.
    -   Unit tests added to validate configuration options, similar to `TableOptions` tests in DataFusion.
    
-   **DataFrameReader:**
    
    -   Enabled reading DataFrames from `csv`, `json`, `orc`, `parquet` and `text` formats.
    -   Supported predefined reading options using the `ConfigOpts` trait for flexible configuration.
    -   Added advanced options like schema merging and recursive file lookup for Parquet files.
    -   Developed tests ensuring correct functionality of the `DataFrameReader` configurations and file parsing.

# Related Issue(s)

-   Closes #54 

# Documentation

-   [PySpark IO](https://spark.apache.org/docs/latest/api/python/reference/pyspark.sql/io.html)
-   [DataFusion](https://docs.rs/datafusion/latest/datafusion/common/config/index.html)

### CODE REVIEW

File: core/src/readwriter.rs [hunk 1]
```diff
@@ -39,6 +39,732 @@ impl ToSchema for &str {
...
+/// let options = JsonOptions::new()
+///     .schema("name STRING, salary INT")
+///     .multi_line(true)
+///     .allow_comments(true)
+///     .encoding("UTF-8".to_string());
+///
+/// let df = spark.read().json("/path/to/json", options)?;
+/// ```
+#[derive(Debug)]
+pub struct JsonOptions {
```
Comment (sjrusso8) [inline core/src/readwriter.rs:146]: Could you also add in these options?

- `time_zone`
- `timestamp_ntz_format`
- `enable_datetime_parsing_fallback`
- `compression`
- `ignore_null_fields`

https://spark.apache.org/docs/latest/sql-data-sources-json.html

File: core/src/readwriter.rs [hunk 2]
```diff
@@ -39,6 +39,732 @@ impl ToSchema for &str {
...
+/// ```
+/// let options = OrcOptions::new()
+///     .merge_schema(true)
+///     .path_glob_filter("*.orc".to_string())
+///     .recursive_file_lookup(true);
+///
+/// let df = spark.read().orc("/path/to/orc", options)?;
+/// ```
+#[derive(Debug)]
+pub struct OrcOptions {
```
Comment (sjrusso8) [inline core/src/readwriter.rs:458]: can you also add in the `compression` option?

https://spark.apache.org/docs/latest/sql-data-sources-orc.html

File: core/src/readwriter.rs [hunk 3]
```diff
@@ -39,6 +39,732 @@ impl ToSchema for &str {
...
+/// ```
+/// let options = ParquetOptions::new()
+///     .merge_schema(true)
+///     .path_glob_filter("*.parquet".to_string())
+///     .recursive_file_lookup(true);
+///
+/// let df = spark.read().parquet("/path/to/parquet", options)?;
+/// ```
+#[derive(Debug)]
+pub struct ParquetOptions {
```
Comment (sjrusso8) [inline core/src/readwriter.rs:556]: can you also add the `compression` option?

File: core/src/readwriter.rs
Comment (sjrusso8) [inline core/src/readwriter.rs:1]: Spark is really specific about casing. I just did a test of the using  `infer_schema` vs `inferSchema` on the CsvOptions, and if i set the option with a key of `infer_schema` Spark did not use the value. It did would work when i set it with `inferSchema`.

Using the below settings, I would expect the DataFrame schema to be `String`, `Integer`, `String`. 

image

But the output I get is String, String, String. I expected the column `age` to be an integer.

image

When I changed the option under the `to_options` to be the `inferSchema` as the key, it worked.

**Review by sjrusso8 — COMMENTED** [review]
Overall this looks good! Left some comments for your consideration. I think the key values when creating the `HashMap` need to be the `camelCase` option from the spark docs.

File: core/src/readwriter.rs [hunk 4]
```diff
@@ -39,6 +39,732 @@ impl ToSchema for &str {
...
+
+/// A struct that represents options for configuring
+/// CsvOptions for CSV file parsing
+pub struct CsvOptions {
+    pub header: Option<bool>,
+    pub delimiter: Option<u8>,
+    pub null_value: Option<String>,
+}
+
+impl CsvOptions {
```
[the remark this answers is not in the document]
Comment (lexara-prime-ai) [reply core/src/readwriter.rs:58]: You're right. I think I missed that. Adding them right away!

File: core/src/readwriter.rs
[the remark this answers is not in the document]
Comment (lexara-prime-ai) [reply core/src/readwriter.rs:1]: I was actually debating about whether to derive Clone for each struct. I also noticed I missed `#[derive(Debug)]`  for CsvOptions as well.

File: core/src/readwriter.rs [hunk 5]
```diff
@@ -39,6 +39,732 @@ impl ToSchema for &str {
...
+/// ```
+#[derive(Debug)]
+pub struct JsonOptions {
+    pub schema: Option<String>,
+    pub primitives_as_string: Option<bool>,
+    pub prefers_decimal: Option<bool>,
+    pub allow_comments: Option<bool>,
+    pub allow_unquoted_field_names: Option<bool>,
+    pub allow_single_quotes: Option<bool>,
+    pub allow_numeric_leading_zero: Option<bool>,
```
[the remark this answers is not in the document]
Comment (lexara-prime-ai) [reply core/src/readwriter.rs:153]: Nice catch! Going through the file changes now as I make the necessary updates.

File: core/src/readwriter.rs (hunk 1 above)
Comment (lexara-prime-ai) [reply core/src/readwriter.rs:146]: Right away!

File: core/src/readwriter.rs
Comment (lexara-prime-ai) [reply core/src/readwriter.rs:1]: I just ran the test after making some updates, the schema looks correct now.

```bash
...
                            name: "name",
                            data_type: Some(
                                DataType {
                                    kind: Some(
                                        String(
                                            String {
                                                type_variation_reference: 0,
                                            },
                                        ),
                                    ),
                                },
                            ),
                            nullable: true,
                            metadata: None,
                        },
                        StructField {
                            name: "age",
                            data_type: Some(
                                DataType {
                                    kind: Some(
                                        Integer(
                                            Integer {
                                                type_variation_reference: 0,
                                            },
                                        ),
                                    ),
                                },
                            ),
                            nullable: true,
                            metadata: None,
                        },
                        StructField {
...
)
test readwriter::tests::test_dataframe_read_csv_with_options ... ok
```

File: core/src/readwriter.rs [hunk 6]
```diff
@@ -39,6 +39,732 @@ impl ToSchema for &str {
...
+/// and bulk options via [options]
+///
+/// This sets multiple options at once using a HashMap
+pub trait ConfigOpts {
+    fn to_options(&self) -> HashMap<String, String>;
+}
+
+/// A struct that represents options for configuring
+/// CsvOptions for CSV file parsing
+pub struct CsvOptions {
```
[the remark this answers is not in the document]
Comment (lexara-prime-ai) [reply core/src/readwriter.rs:162]: Just updated the option field to expect a `String`. Hopefully, this'll provide some flexibility.

File: core/src/readwriter.rs (hunk 3 above)
Comment (lexara-prime-ai) [reply core/src/readwriter.rs:1067]: Just added the `compression` for the `Option` structs.

File: core/src/readwriter.rs [hunk 7]
```diff
@@ -39,6 +39,732 @@ impl ToSchema for &str {
     }
 }
 
+/// A trait used to handle individual options via [option]
+/// and bulk options via [options]
+///
+/// This sets multiple options at once using a HashMap
+pub trait ConfigOpts {
```
[the remark this answers is not in the document]
Comment (lexara-prime-ai) [reply core/src/readwriter.rs:46]: We've added a `CommonFileOptions` struct and implemented `ConfigOpts` for `CommonFileOptions`. It reduced the amount of boilerplate, especially for the `ConfigOpts` implementations for each `Options` struct. Looking forward to some feedback concerning this approach @sjrusso8 .

File: core/src/readwriter.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (sjrusso8) [reply core/src/readwriter.rs]: @lexara-prime-ai  I think this might be the last change :)

[the remark this answers is not in the document]
Comment (lexara-prime-ai) [reply core/src/readwriter.rs]: Awesome🎊🎊🎊!

**Comment by lexara-prime-ai** [comment]
Hi @sjrusso8  I just updated the pr.

**Comment by sjrusso8** [comment]
@lexara-prime-ai  LGTM! just update the README.md and mark these as closed. Then i'll merge in the change

**Review by sjrusso8 — APPROVED** [review]

**Pull request merged**